components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m

```diff
@@ -14,89 +14,527 @@
 
 #import <XCTest/XCTest.h>
 
+#import <objc/runtime.h>
 #import "MaterialTextFields+ContainedInputView.h"
 
-@interface MDCBaseTextField (Testing)
+@interface MDCBaseTextField (Private)
 @property(nonatomic, assign) UIUserInterfaceLayoutDirection layoutDirection;
+- (CGRect)adjustTextAreaFrame:(CGRect)textRect
+    withParentClassTextAreaFrame:(CGRect)parentClassTextAreaFrame;
+@end
+
+@interface MDCBaseTextField (Testing)
+@property(nonatomic, assign) BOOL isEditingOverride;
+@end
+
+@implementation MDCBaseTextField (Testing)
+@dynamic isEditingOverride;
+- (BOOL)isEditingOverride {
+  return [objc_getAssociatedObject(self, _cmd) boolValue];
+}
+- (void)setIsEditingOverride:(BOOL)isEditingOverride {
+  objc_setAssociatedObject(self, @selector(isEditingOverride), @(isEditingOverride),
+                           OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+- (BOOL)isEditing {
+  return self.isEditingOverride ? self.isEditingOverride : [super isEditing];
+}
 @end
 
 @interface MDCBaseTextFieldTests : XCTestCase
 @end
 
 @implementation MDCBaseTextFieldTests
 
+#pragma mark Helper Methods
+
 - (UIView *)createSideView {
   UIView *sideView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 20, 20)];
   sideView.backgroundColor = [UIColor blueColor];
   return sideView;
 }
 
-- (void)testLeadingViewLTR {
+- (MDCBaseTextField *)createLTRTextFieldWithLeadingView:(BOOL)leadingView
+                                           trailingView:(BOOL)trailingView
+                                               viewMode:(UITextFieldViewMode)viewMode {
+  CGRect frame = CGRectMake(0, 0, 100, 60);
+  MDCBaseTextField *textField = [[MDCBaseTextField alloc] initWithFrame:frame];
+  textField.layoutDirection = UIUserInterfaceLayoutDirectionLeftToRight;
+  if (leadingView) {
+    textField.leadingView = [self createSideView];
+  }
+  if (trailingView) {
+    textField.trailingView = [self createSideView];
+  }
+  textField.trailingViewMode = viewMode;
+  textField.leadingViewMode = viewMode;
+  UIWindow *window = [[UIWindow alloc] init];
```

Comment:
Is it possible to return a text field that has not been added to a UIWindow? If not, please document why we need the UIWindow here.

Reply:
None of the tests passed locally until I added them to a window :/ . Still can't get them to pass on kokoro, but that's another story.